lib/xs_transport.ml

```diff
@@ -13,3 +13,23 @@
  *)
 
 let xenstored_socket = ref "/var/run/xenstored/socket"
+
+(* We'll look for these paths in order: *)
+let get_xenstore_paths () =
+  let default = [
+    !xenstored_socket;
+    "/proc/xen/xenbus"; (* Linux *)
+    "/dev/xen/xenstore"; (* FreeBSD *)
+  ] in
+  try
+    Sys.getenv "XENSTORED_PATH" :: default
+  with Not_found -> default
+
+let choose_xenstore_path () =
+  List.fold_left (fun acc possibility -> match acc with
+    | Some x -> Some x
+    | None ->
+      if Sys.file_exists possibility then Some possibility else None
+  ) None (get_xenstore_paths ())
```

Comment:
I think this could be done with `List.find`, as that also returns the first element that satisfies our predicate:
```ocaml
try Some(List.find Sys.file_exists (get_xenstore_paths ()))
with Not_found -> None
```